File: packages/command-client/src/tauri/ssh.ts
import { Channel, invoke } from "@tauri-apps/api/core";
import { call, isTauriRuntime } from "./invoke";
import type {
  SshCloseInput,
  SshConnectInput,
  SshConnection,
  SshConnectionInput,
  SshHostFingerprintInfo,
  SshHostKeyInput,
  SshKnownHostsExportInput,
  SshKnownHostsExportResult,
  SshKnownHostsImportInput,
  SshKnownHostsImportResult,
  SshLogExport,
  SshLogExportInput,
  SshReconnectCancelInput,
  SshResizeInput,
  SshSessionEvent,
  SshSessionInput,
  SshSessionSummary,
  SshTestResult,
  SshTask,
  SshTaskCancelInput,
  SshTaskCleanupInput,
  SshTaskCleanupResult,
  SshTaskDetail,
  SshTaskRun,
  SshTaskRunEvent,
  SshTaskRunInput,
  SshTaskSaveInput,
  SftpCancelTransferInput,
  SftpDeleteInput,
  SftpDirectoryListing,
  SftpFileEntry,
  SftpOpenResult,
  SftpPathInput,
  SftpRenameInput,
  SftpSessionInput,
  SftpTransferInput,
  SftpTransferState,
} from "../types";

export function listSshTasks(workspaceId: string) {
  return call<SshTask[]>("ssh_tasks_list", { workspaceId });
}

export function getSshTask(workspaceId: string, taskId: string) {
  return call<SshTaskDetail>("ssh_task_get", { workspaceId, taskId });
}

export function saveSshTask(input: SshTaskSaveInput) {
  return call<SshTaskDetail>("ssh_task_save", { input });
}

export function duplicateSshTask(workspaceId: string, taskId: string) {
  return call<SshTaskDetail>("ssh_task_duplicate", { workspaceId, taskId });
}

export function deleteSshTask(workspaceId: string, taskId: string) {
  return call<void>("ssh_task_delete", { workspaceId, taskId });
}

export function runSshTask(input: SshTaskRunInput) {
  return call<SshTaskRun>("ssh_task_run", { input });
}

export function cancelSshTaskRun(input: SshTaskCancelInput) {
  return call<SshTaskRun>("ssh_task_run_cancel", { input });
}

export function listSshTaskRuns(workspaceId: string, taskId: string) {
  return call<SshTaskRun[]>("ssh_task_runs_list", { workspaceId, taskId });
}

export function readSshTaskRunLog(workspaceId: string, runId: string) {
  return call<string>("ssh_task_run_log_read", { workspaceId, runId });
}

export function clearSshTaskRuns(input: SshTaskCleanupInput) {
  return call<SshTaskCleanupResult>("ssh_task_runs_clear", { input });
}

export function listSshConnections(workspaceId: string) {
  return call<SshConnection[]>("ssh_connections_list", { workspaceId });
}

export function saveSshConnection(input: SshConnectionInput) {
  return call<SshConnection>("ssh_connection_save", { input });
}

export function testSshConnection(input: SshConnectionInput) {
  return call<SshTestResult>("ssh_connection_test", { input });
}

export function deleteSshConnection(workspaceId: string, connectionId: string) {
  return call<SshConnection[]>("ssh_connection_delete", {
    workspaceId,
    connectionId,
  });
}

export function connectSshSession(input: SshConnectInput) {
  return call<SshSessionSummary>("ssh_session_connect", { input });
}

export function listSshSessions(workspaceId: string) {
  return call<SshSessionSummary[]>("ssh_sessions_list", { workspaceId });
}

export function getSshSessionHistory(input: SshCloseInput) {
  return call<SshSessionEvent[]>("ssh_session_history", { input });
}

export function sendSshInput(input: SshSessionInput) {
  return call<SshSessionEvent>("ssh_session_input", { input });
}

export function resizeSshSession(input: SshResizeInput) {
  return call<SshSessionEvent>("ssh_session_resize", { input });
}

export function closeSshSession(input: SshCloseInput) {
  return call<SshSessionSummary>("ssh_session_close", { input });
}

export function cancelSshReconnect(input: SshReconnectCancelInput) {
  return call<SshSessionSummary>("ssh_session_reconnect_cancel", { input });
}

export function exportSshLog(input: SshLogExportInput) {
  return call<SshLogExport>("ssh_session_log_export", { input });
}

export function getSshHostFingerprint(input: SshHostKeyInput) {
  return call<SshHostFingerprintInfo | null>("ssh_host_key_get", { input });
}

export function resetSshHostFingerprint(input: SshHostKeyInput) {
  return call<boolean>("ssh_host_key_reset", { input });
}

export function listSshHostFingerprints(workspaceId: string) {
  return call<SshHostFingerprintInfo[]>("ssh_host_key_list", { workspaceId });
}

export function importSshKnownHosts(input: SshKnownHostsImportInput) {
  return call<SshKnownHostsImportResult>("ssh_known_hosts_import", { input });
}

export function exportSshKnownHosts(input: SshKnownHostsExportInput) {
  return call<SshKnownHostsExportResult>("ssh_known_hosts_export", { input });
}

export function openSftp(input: SftpSessionInput) {
  return call<SftpOpenResult>("ssh_sftp_open", { input });
}

export function listSftpDirectory(input: SftpPathInput) {
  return call<SftpDirectoryListing>("ssh_sftp_list_directory", { input });
}

export function statSftpPath(input: SftpPathInput) {
  return call<SftpFileEntry>("ssh_sftp_stat", { input });
}

export function createSftpDirectory(input: SftpPathInput) {
  return call<void>("ssh_sftp_create_directory", { input });
}

export function renameSftpPath(input: SftpRenameInput) {
  return call<void>("ssh_sftp_rename", { input });
}

export function deleteSftpPath(input: SftpDeleteInput) {
  return call<void>("ssh_sftp_delete", { input });
}

export function downloadSftpFile(input: SftpTransferInput) {
  return call<SftpTransferState>("ssh_sftp_download", { input });
}

export function uploadSftpFile(input: SftpTransferInput) {
  return call<SftpTransferState>("ssh_sftp_upload", { input });
}

export function cancelSftpTransfer(input: SftpCancelTransferInput) {
  return call<SftpTransferState>("ssh_sftp_cancel_transfer", { input });
}

export function listSftpTransfers(input: SftpSessionInput) {
  return call<SftpTransferState[]>("ssh_sftp_transfers_list", { input });
}

export type SshTerminalDataPayload = {
  sessionId: string;
  data: string;
  status?: SshSessionSummary["status"] | null;
  reconnectAttempt?: number;
};

/**
 * Subscribe to live SSH terminal output over a Tauri IPC `Channel` (the same
 * reliable transport used by commands) rather than the event system. High-rate
 * Tauri events stall WebView2 event delivery on Windows under a full-screen
 * redraw burst (vim/less/top), which silently freezes the terminal; channels
 * do not have that failure mode. Returns a disposer that detaches the handler.
 * A no-op outside the Tauri runtime (browser mock mode polls history instead).
 */
export function registerSshTerminalChannel(
  onMessage: (payload: SshTerminalDataPayload) => void,
): Promise<() => void> {
  if (!isTauriRuntime()) {
    return Promise.resolve(() => {});
  }
  const channel = new Channel<SshTerminalDataPayload>();
  channel.onmessage = onMessage;
  return invoke<void>("ssh_register_terminal_channel", { channel })
    .then(() => () => {
      channel.onmessage = () => {};
    })
    .catch(() => () => {
      channel.onmessage = () => {};
    });
}

export function registerSftpTransferChannel(
  onMessage: (payload: SftpTransferState) => void,
): Promise<() => void> {
  if (!isTauriRuntime()) {
    return Promise.resolve(() => {});
  }
  const channel = new Channel<SftpTransferState>();
  channel.onmessage = onMessage;
  return invoke<void>("ssh_register_sftp_transfer_channel", { channel })
    .then(() => () => {
      channel.onmessage = () => {};
    })
    .catch(() => () => {
      channel.onmessage = () => {};
    });
}

export function registerSshTaskRunChannel(
  onMessage: (payload: SshTaskRunEvent) => void,
): Promise<() => void> {
  if (!isTauriRuntime()) {
    return Promise.resolve(() => {});
  }
  const channel = new Channel<SshTaskRunEvent>();
  channel.onmessage = onMessage;
  return invoke<void>("ssh_register_task_run_channel", { channel })
    .then(() => () => {
      channel.onmessage = () => {};
    })
    .catch(() => () => {
      channel.onmessage = () => {};
    });
}

File: crates/unfour-app/src/commands/ssh.rs
use crate::AppState;
use tauri::State;
use unfour_core::{
    models::{
        SftpCancelTransferInput, SftpDeleteInput, SftpDirectoryListing, SftpFileEntry,
        SftpOpenResult, SftpPathInput, SftpRenameInput, SftpSessionInput, SftpTransferInput,
        SftpTransferState, SshCloseInput, SshConnectInput, SshConnection, SshConnectionInput,
        SshHostFingerprintInfo, SshHostKeyInput, SshKnownHostsExportInput,
        SshKnownHostsExportResult, SshKnownHostsImportInput, SshKnownHostsImportResult,
        SshLogExport, SshLogExportInput, SshReconnectCancelInput, SshResizeInput, SshSessionEvent,
        SshSessionInput, SshSessionSummary, SshTask, SshTaskCancelInput, SshTaskCleanupInput,
        SshTaskCleanupResult, SshTaskDetail, SshTaskRun, SshTaskRunInput, SshTaskSaveInput,
        SshTestResult,
    },
    AppResult,
};

use super::trace_command;

#[tauri::command]
pub async fn ssh_tasks_list(
    workspace_id: String,
    state: State<'_, AppState>,
) -> AppResult<Vec<SshTask>> {
    state.command_bus.list_ssh_tasks(workspace_id).await
}

#[tauri::command]
pub async fn ssh_task_get(
    workspace_id: String,
    task_id: String,
    state: State<'_, AppState>,
) -> AppResult<SshTaskDetail> {
    state.command_bus.get_ssh_task(workspace_id, task_id).await
}

#[tauri::command]
pub async fn ssh_task_save(
    input: SshTaskSaveInput,
    state: State<'_, AppState>,
) -> AppResult<SshTaskDetail> {
    state.command_bus.save_ssh_task(input).await
}

#[tauri::command]
pub async fn ssh_task_duplicate(
    workspace_id: String,
    task_id: String,
    state: State<'_, AppState>,
) -> AppResult<SshTaskDetail> {
    state
        .command_bus
        .duplicate_ssh_task(workspace_id, task_id)
        .await
}

#[tauri::command]
pub async fn ssh_task_delete(
    workspace_id: String,
    task_id: String,
    state: State<'_, AppState>,
) -> AppResult<()> {
    state
        .command_bus
        .delete_ssh_task(workspace_id, task_id)
        .await
}

#[tauri::command]
pub async fn ssh_task_run(
    input: SshTaskRunInput,
    state: State<'_, AppState>,
) -> AppResult<SshTaskRun> {
    trace_command("ssh_task_run", state.command_bus.run_ssh_task(input)).await
}

#[tauri::command]
pub async fn ssh_task_run_cancel(
    input: SshTaskCancelInput,
    state: State<'_, AppState>,
) -> AppResult<SshTaskRun> {
    state.command_bus.cancel_ssh_task_run(input).await
}

#[tauri::command]
pub async fn ssh_task_runs_list(
    workspace_id: String,
    task_id: String,
    state: State<'_, AppState>,
) -> AppResult<Vec<SshTaskRun>> {
    state
        .command_bus
        .list_ssh_task_runs(workspace_id, task_id)
        .await
}

#[tauri::command]
pub async fn ssh_task_run_log_read(
    workspace_id: String,
    run_id: String,
    state: State<'_, AppState>,
) -> AppResult<String> {
    state
        .command_bus
        .read_ssh_task_run_log(workspace_id, run_id)
        .await
}

#[tauri::command]
pub async fn ssh_task_runs_clear(
    input: SshTaskCleanupInput,
    state: State<'_, AppState>,
) -> AppResult<SshTaskCleanupResult> {
    state.command_bus.clear_ssh_task_runs(input).await
}

#[tauri::command]
pub async fn ssh_register_task_run_channel(
    channel: tauri::ipc::Channel<serde_json::Value>,
    state: State<'_, AppState>,
) -> AppResult<()> {
    if let Ok(mut guard) = state.task_run_channel.lock() {
        *guard = Some(channel);
    }
    Ok(())
}

#[tauri::command]
pub async fn ssh_connections_list(
    workspace_id: String,
    state: State<'_, AppState>,
) -> AppResult<Vec<SshConnection>> {
    state.command_bus.list_ssh_connections(workspace_id).await
}

#[tauri::command]
pub async fn ssh_connection_save(
    input: SshConnectionInput,
    state: State<'_, AppState>,
) -> AppResult<SshConnection> {
    state.command_bus.save_ssh_connection(input).await
}

#[tauri::command]
pub async fn ssh_connection_test(
    input: SshConnectionInput,
    state: State<'_, AppState>,
) -> AppResult<SshTestResult> {
    state.command_bus.test_ssh_connection(input).await
}

#[tauri::command]
pub async fn ssh_connection_delete(
    workspace_id: String,
    connection_id: String,
    state: State<'_, AppState>,
) -> AppResult<Vec<SshConnection>> {
    state
        .command_bus
        .delete_ssh_connection(workspace_id, connection_id)
        .await
}

#[tauri::command]
pub async fn ssh_session_connect(
    input: SshConnectInput,
    state: State<'_, AppState>,
) -> AppResult<SshSessionSummary> {
    trace_command(
        "ssh_session_connect",
        state.command_bus.connect_ssh_session(input),
    )
    .await
}

#[tauri::command]
pub async fn ssh_sessions_list(
    workspace_id: String,
    state: State<'_, AppState>,
) -> AppResult<Vec<SshSessionSummary>> {
    state.command_bus.list_ssh_sessions(workspace_id).await
}

#[tauri::command]
pub async fn ssh_session_history(
    input: SshCloseInput,
    state: State<'_, AppState>,
) -> AppResult<Vec<SshSessionEvent>> {
    state.command_bus.ssh_session_history(input).await
}

#[tauri::command]
pub async fn ssh_session_input(
    input: SshSessionInput,
    state: State<'_, AppState>,
) -> AppResult<SshSessionEvent> {
    state.command_bus.send_ssh_input(input).await
}

/// Register the frontend's IPC channel as the live terminal-output sink. Live
/// SSH output then streams over this channel instead of the Tauri event system,
/// which stalls under the high-rate emit burst of a full-screen redraw on
/// WebView2/Windows. The most recently registered channel wins.
#[tauri::command]
pub async fn ssh_register_terminal_channel(
    channel: tauri::ipc::Channel<serde_json::Value>,
    state: State<'_, AppState>,
) -> AppResult<()> {
    if let Ok(mut guard) = state.terminal_channel.lock() {
        *guard = Some(channel);
    }
    Ok(())
}

#[tauri::command]
pub async fn ssh_register_sftp_transfer_channel(
    channel: tauri::ipc::Channel<serde_json::Value>,
    state: State<'_, AppState>,
) -> AppResult<()> {
    if let Ok(mut guard) = state.sftp_transfer_channel.lock() {
        *guard = Some(channel);
    }
    Ok(())
}

#[tauri::command]
pub async fn ssh_sftp_open(
    input: SftpSessionInput,
    state: State<'_, AppState>,
) -> AppResult<SftpOpenResult> {
    trace_command("ssh_sftp_open", state.command_bus.open_ssh_sftp(input)).await
}

#[tauri::command]
pub async fn ssh_sftp_list_directory(
    input: SftpPathInput,
    state: State<'_, AppState>,
) -> AppResult<SftpDirectoryListing> {
    state.command_bus.list_ssh_sftp_directory(input).await
}

#[tauri::command]
pub async fn ssh_sftp_stat(
    input: SftpPathInput,
    state: State<'_, AppState>,
) -> AppResult<SftpFileEntry> {
    state.command_bus.stat_ssh_sftp_path(input).await
}

#[tauri::command]
pub async fn ssh_sftp_create_directory(
    input: SftpPathInput,
    state: State<'_, AppState>,
) -> AppResult<()> {
    state.command_bus.create_ssh_sftp_directory(input).await
}

#[tauri::command]
pub async fn ssh_sftp_rename(input: SftpRenameInput, state: State<'_, AppState>) -> AppResult<()> {
    state.command_bus.rename_ssh_sftp_path(input).await
}

#[tauri::command]
pub async fn ssh_sftp_delete(input: SftpDeleteInput, state: State<'_, AppState>) -> AppResult<()> {
    state.command_bus.delete_ssh_sftp_path(input).await
}

#[tauri::command]
pub async fn ssh_sftp_download(
    input: SftpTransferInput,
    state: State<'_, AppState>,
) -> AppResult<SftpTransferState> {
    state.command_bus.download_ssh_sftp_file(input).await
}

#[tauri::command]
pub async fn ssh_sftp_upload(
    input: SftpTransferInput,
    state: State<'_, AppState>,
) -> AppResult<SftpTransferState> {
    state.command_bus.upload_ssh_sftp_file(input).await
}

#[tauri::command]
pub async fn ssh_sftp_cancel_transfer(
    input: SftpCancelTransferInput,
    state: State<'_, AppState>,
) -> AppResult<SftpTransferState> {
    state.command_bus.cancel_ssh_sftp_transfer(input).await
}

#[tauri::command]
pub async fn ssh_sftp_transfers_list(
    input: SftpSessionInput,
    state: State<'_, AppState>,
) -> AppResult<Vec<SftpTransferState>> {
    state.command_bus.list_ssh_sftp_transfers(input).await
}

#[tauri::command]
pub async fn ssh_session_resize(
    input: SshResizeInput,
    state: State<'_, AppState>,
) -> AppResult<SshSessionEvent> {
    trace_command(
        "ssh_session_resize",
        state.command_bus.resize_ssh_session(input),
    )
    .await
}

#[tauri::command]
pub async fn ssh_session_close(
    input: SshCloseInput,
    state: State<'_, AppState>,
) -> AppResult<SshSessionSummary> {
    trace_command(
        "ssh_session_close",
        state.command_bus.close_ssh_session(input),
    )
    .await
}

#[tauri::command]
pub async fn ssh_session_reconnect_cancel(
    input: SshReconnectCancelInput,
    state: State<'_, AppState>,
) -> AppResult<SshSessionSummary> {
    state.command_bus.cancel_ssh_reconnect(input).await
}

#[tauri::command]
pub async fn ssh_session_log_export(
    input: SshLogExportInput,
    state: State<'_, AppState>,
) -> AppResult<SshLogExport> {
    state.command_bus.export_ssh_log(input).await
}

#[tauri::command]
pub async fn ssh_host_key_get(
    input: SshHostKeyInput,
    state: State<'_, AppState>,
) -> AppResult<Option<SshHostFingerprintInfo>> {
    state.command_bus.get_ssh_host_fingerprint(input).await
}

#[tauri::command]
pub async fn ssh_host_key_reset(
    input: SshHostKeyInput,
    state: State<'_, AppState>,
) -> AppResult<bool> {
    state.command_bus.reset_ssh_host_fingerprint(input).await
}

#[tauri::command]
pub async fn ssh_host_key_list(
    workspace_id: String,
    state: State<'_, AppState>,
) -> AppResult<Vec<SshHostFingerprintInfo>> {
    state
        .command_bus
        .list_all_ssh_fingerprints(workspace_id)
        .await
}

#[tauri::command]
pub async fn ssh_known_hosts_import(
    input: SshKnownHostsImportInput,
    state: State<'_, AppState>,
) -> AppResult<SshKnownHostsImportResult> {
    state.command_bus.import_ssh_known_hosts(input).await
}

#[tauri::command]
pub async fn ssh_known_hosts_export(
    input: SshKnownHostsExportInput,
    state: State<'_, AppState>,
) -> AppResult<SshKnownHostsExportResult> {
    state.command_bus.export_ssh_known_hosts(input).await
}
